server/ec2spotmanager/tasks.py

```diff
@@ -131,566 +113,247 @@ def check_instance_pool(pool_id):
+        prices[instance_type] = json.loads(data)
+    
+    for instance_type in prices.keys():
+        for region in prices[instance_type]:
+            for zone in prices[instance_type][region]:
+                if cache.get('ec2spot:blacklist:%s:%s' % (zone, instance_type)) is not None:
+                    blacklisted_zones.append((zone, instance_type))
 
-    for region in instance_ids_by_region:
-        cluster = EC2Manager(None)
-        try:
-            cluster.connect(region=region, aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
-                            aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY)
-        except Exception as msg:
-            # Log this error to the pool status messages
-            entry = PoolStatusEntry()
-            entry.type = POOL_STATUS_ENTRY_TYPE['unclassified']
-            entry.pool = pool
-            entry.msg = str(msg)
-            entry.isCritical = True
-            entry.save()
+    return (prices, blacklisted_zones)
 
-            logger.exception("[Pool %d] terminate_pool_instances: laniakea failure: %s", pool.id, msg)
-            return None
 
-        try:
-            if terminateByPool:
-                boto_instances = cluster.find(filters={"tag:" + SPOTMGR_TAG + "-PoolId": str(pool.pk)})
-
-                # Data consistency checks
-                for boto_instance in boto_instances:
-                    # state_code is a 16-bit value where the high byte is
-                    # an opaque internal value and should be ignored.
-                    state_code = boto_instance.state_code & 255
-                    if not ((boto_instance.id in instance_ids_by_region[region]) or
-                            (state_code == INSTANCE_STATE['shutting-down'] or
-                             state_code == INSTANCE_STATE['terminated'])):
-                        logger.error("[Pool %d] Instance with EC2 ID %s (status %d) "
-                                     "is not in region list for region %s",
-                                     pool.id, boto_instance.id, state_code, region)
-
-                cluster.terminate(boto_instances)
-            else:
-                logger.info("[Pool %d] Terminating %s instances in region %s",
-                            pool.id, len(instance_ids_by_region[region]), region)
-                cluster.terminate(cluster.find(instance_ids=instance_ids_by_region[region]))
-        except (boto.exception.EC2ResponseError, boto.exception.BotoServerError, ssl.SSLError, socket.error) as msg:
-            logger.exception("[Pool %d] terminate_pool_instances: boto failure: %s", pool.id, msg)
-            return 1
 
+def _terminate_pool_instances(instance_pool, running_instances, terminateByPool=False):
+    """ Terminate an instance with the given configuration """
+    cloud_provider = Provider().getInstance(instance_pool.config.provider)
+    
+    try:
+        cloud_provider.terminate_instances(instance_pool, running_instances, terminateByPool=True)
```

Comment:
`instance_pool` and `running_instances` are both Django models, and shouldn't be passed to the abstraction directly.

Reply:
this has been changed. this only takes instance ids.